pkg/lib/repo/local/registry.go

```diff
@@ -0,0 +1,183 @@
+package local
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"slices"
+	"strings"
+
+	"kitops/pkg/lib/constants"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/content/oci"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/registry"
+)
+
+type LocalRepo interface {
+	GetRepoName() string
+	BlobPath(ocispec.Descriptor) string
+	GetAllModels() []ocispec.Descriptor
+	GetTags(ocispec.Descriptor) []string
+	oras.Target
+	content.Deleter
+	content.Untagger
+}
+
+type localRepo struct {
+	storagePath string
+	nameRef     string
+	localIndex  *localIndex
+	*oci.Store
+}
+
+func NewLocalRepo(storagePath string, ref *registry.Reference) (LocalRepo, error) {
+	nameRef := path.Join(ref.Registry, ref.Repository)
+	return newLocalRepoForName(storagePath, nameRef)
+}
+
+func newLocalRepoForName(storagePath, name string) (LocalRepo, error) {
+	repo := &localRepo{}
+	repo.storagePath = storagePath
+	repo.nameRef = name
+
+	store, err := oci.New(storagePath)
+	if err != nil {
+		return nil, err
+	}
+	repo.Store = store
+
+	// Initialize repo-specific index.json
+	localIndex, err := newLocalIndex(storagePath, name)
+	if err != nil {
+		return nil, err
+	}
+	repo.localIndex = localIndex
+
+	return repo, nil
+}
+
+func GetAllLocalRepos(storagePath string) ([]LocalRepo, error) {
+	entries, err := os.ReadDir(storagePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read local storage: %w", err)
+	}
+
+	var repos []LocalRepo
+	for _, dirEntry := range entries {
+		if dirEntry.IsDir() {
+			continue
+		}
+		if !constants.FileIsLocalIndex(dirEntry.Name()) {
+			continue
+		}
+		repoName, err := constants.RepoForIndexJsonPath(dirEntry.Name())
+		if err != nil {
+			return nil, err
+		}
+		repo, err := newLocalRepoForName(storagePath, repoName)
+		if err != nil {
+			return nil, err
+		}
+		repos = append(repos, repo)
+	}
+
+	// Sort alphabetically
+	slices.SortFunc(repos, func(a, b LocalRepo) int {
+		return strings.Compare(a.GetRepoName(), b.GetRepoName())
+	})
+
+	return repos, nil
+}
+
+// GetRepo returns the registry and repository for the current OCI store.
+func (r *localRepo) GetRepoName() string {
+	return r.nameRef
+}
+
+func (r *localRepo) BlobPath(desc ocispec.Descriptor) string {
+	return filepath.Join(r.storagePath, "blobs", "sha256", desc.Digest.Encoded())
+}
+
+func (l *localRepo) Delete(ctx context.Context, target ocispec.Descriptor) error {
+	if target.MediaType != ocispec.MediaTypeImageManifest {
+		return l.Store.Delete(ctx, target)
+	}
+
+	canDelete, err := canSafelyDeleteManifest(ctx, l.storagePath, target)
+	if err != nil {
+		return err
```

Comment:
Added context to this error, but I'm not sure we need more on the other ones -- they'll generally have os/file writing errors. We want to avoid logging too much in this handler code because the user shouldn't care about the idiosyncrasies of how we store files and why we can't _actually_ remove a manifest yet.